### PR TITLE
NEPT-1999: Schedule publishing of new version of node

### PR DESCRIPTION
**Pull request by enriquelacoma**
### Description

Users need to be able to schedule the publishing of a new version of an already published node.
### Change log

- Added: add patches to the deploy 
    .- patches/scheduler_workbench-allowed_status.patch
    .- https://www.drupal.org/files/issues/scheduler_workbench-revision_publish-2048999-45.patch

### CODE REVIEW

File: resources/multisite_drupal_standard.make
Comment (imanoleguskiza) [inline resources/multisite_drupal_standard.make]: There is a type here, it should say "This patch requires the patch from issue "

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: There is a type here, it should say "When Scheduler Workbench tries to publish a revision **it** will check"

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: The permission name here differs from the one defined below: here its "administer scheduler workbench" but then we defined it as "scheduler workbench"

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: The functione name here is a bit weird, it doesn't tell anything about what it does. I think a better name would be _scheduler_workbench_define_allowed_status

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: This interaction seems unnecessary, can't we simply add $value->from_name to the $status array if the condition is met?
`if ($value->to_name == 'published') {
  $status[] = $value->from_name;
}
`

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: The variable name is misleading here, I think $allowed_status is much better ( same for other appearances of the same variable)

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: In which situation will value equal 0?

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: The code added above should probable be inside this condition, that way we don't do unnecessary checks if the content type doesn't have scheduler enabled.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch]: The check here is fine but I think we shouldn't allow the users to define a scheduling date if the revision state is not allowed to begin with.
I would do this on scheduler_workbench_form_alter, just before calling scheduler_workbench_add_revision_publish_on_field and I would add a message to inform the user that publishing the revision in that state is not allowed.

**Review by imanoleguskiza — CHANGES_REQUESTED** [review]

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: This allows, for example, to schedule the publication of a draft and publish it only if it is in Validation, that way you can schedule a publication and publish it only if it is in a correct status.

If a revision is not in a valid status, i will not be published and register a watchdog message with the error.

I think that we could keep the current funtionality and, as you suggest, add a message informing the user that it will no be published with the current status.

File: tests/src/Context/SchedulerContext.php
Comment (imanoleguskiza) [inline tests/src/Context/SchedulerContext.php:66]: This two steps seem to be unused so I guess we should delete them.

File: tests/features/scheduler.feature
Comment (imanoleguskiza) [inline tests/features/scheduler.feature:41]: The tests are fine but we are missing the tests where we publish a new revision of an existing node. We also need a tests that shows how a node with an incorrect state is not published when the date comes.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (imanoleguskiza) [inline resources/patches/scheduler_workbench-allowed_status.patch:163]: In my tests the watchdog was triggered but the scheduling proceeded as usual so I think the check needs to be triggered somewhere else.

**Review by imanoleguskiza — CHANGES_REQUESTED** [review]

File: tests/src/Context/SchedulerContext.php
Comment (enriquelacoma) [reply tests/src/Context/SchedulerContext.php:66]: We use them now in the scenario to check the publication of revision

File: tests/features/scheduler.feature
Comment (enriquelacoma) [reply tests/features/scheduler.feature:41]: Added two new scenarios

**Review by AlbaParrilla — DISMISSED** [review]

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch]: We should not have this new hook_update in 2.5. I suggest to remove it from the "local" patch.

File: resources/multisite_drupal_standard.make
Comment (gillesdeudon) [inline resources/multisite_drupal_standard.make:677]: We need a reference to the D.O issue from where this "local" patch id derived like that we can ensure the traceability of the patch and ease our job during future module upgrades.

**Review by gillesdeudon — CHANGES_REQUESTED** [review]
@enriquelacoma could you tackle the comments I made?

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch]: IMO, this menu time should be either a child of "Scheduler" item (admin/config/scheduler) or of "Workbench moderation" item (admin/config/workbench/moderation) in order to be consistent with the other contrib modules that enrich Scheduler or Workbench moderation like "Workbench moderation" does with the "Workbench" module. 

I am more in favour of "Scheduler" item with in the `scheduler_workbench_allowed_status_form()`, a mention of the Workbench moderation transitions and the admin form managing them.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch]: IMO, if `scheduler_workbench_allowed_status`is not set, we should come back to the transition settings and only give the list of states that can be lead to "published"; instead of an empty array.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch]: Same comment as I did for `scheduler_workbench_node_validate()`.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch:181]: Same remark as I made for `scheduler_workbench_node_validate()`. Maybe encapsulating this variable_get call in its own function could be interesting.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch:38]: We should tie the options on the Workbench moderation transitions possibilities like done in `_scheduler_workbench_define_allowed_status()`. 
That means this function should move to the module file. I think also it could be integrated in the module API and become public.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch]: In link with my comment above, we could add in the description a text referring to the "Transition" interface of Workbench moderation to add more Moderate state.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: Function removed

File: resources/multisite_drupal_standard.make
Comment (enriquelacoma) [reply resources/multisite_drupal_standard.make:677]: Comment updated

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (gillesdeudon) [inline resources/patches/scheduler_workbench-allowed_status.patch]: Same remark as I made for scheduler_workbench_node_validate(). Maybe encapsulating this variable_get call in its own function could be interesting.

**Review by gillesdeudon — CHANGES_REQUESTED** [review]
I leave some comments to address before being able to merge.

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch:38]: Created wrapper _scheduler_workbench_allowed_status()

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: Updated description to include default values

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: Included as a scheduler tab

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: Created wrapper _scheduler_workbench_allowed_status()

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: Created wrapper _scheduler_workbench_allowed_status()

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch:181]: Created wrapper _scheduler_workbench_allowed_status()

File: resources/patches/scheduler_workbench-allowed_status.patch
Comment (enriquelacoma) [reply resources/patches/scheduler_workbench-allowed_status.patch]: Created wrapper _scheduler_workbench_allowed_status()

**Review by ventzie555 — DISMISSED** [review]
I tested it and everything appears to work fine. Code also look good to me.

**Review by imanoleguskiza — DISMISSED** [review]

**Review by gillesdeudon — DISMISSED** [review]

**Comment by gillesdeudon** [comment]
I approve in order to unblock the PR as it is in a new code review process

**Review by AlbaParrilla — APPROVED** [review]
I just changed a message a little bit, the rest is working properly. 
Ok for me

**Review by Fefaine — APPROVED** [review]
OK

**Comment by Fefaine** [comment]
🎆

**Pull request merged**
